ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/DataColumnSidecarAvailabilityChecker.java

```diff
@@ -0,0 +1,130 @@
+import tech.pegasys.teku.spec.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.spec.datastructures.forkchoice.ReadOnlyStore;
+import tech.pegasys.teku.spec.logic.common.statetransition.availability.AvailabilityChecker;
+import tech.pegasys.teku.spec.logic.common.statetransition.availability.DataAndValidationResult;
+import tech.pegasys.teku.spec.logic.versions.eip7594.helpers.MiscHelpersEip7594;
+import tech.pegasys.teku.statetransition.datacolumns.DataAvailabilitySampler;
+
+public class DataColumnSidecarAvailabilityChecker
+    implements AvailabilityChecker<DataColumnSidecar> {
+  private static final Logger LOG = LogManager.getLogger("das-nyota");
+
+  private final DataAvailabilitySampler dataAvailabilitySampler;
+  private final SafeFuture<DataAndValidationResult<DataColumnSidecar>> validationResult =
+      new SafeFuture<>();
+  final KZG kzg;
+  final Spec spec;
+  final ReadOnlyStore store;
+
+  private final SignedBeaconBlock block;
+
+  public DataColumnSidecarAvailabilityChecker(
+      final DataAvailabilitySampler dataAvailabilitySampler,
+      final ReadOnlyStore store,
+      final KZG kzg,
+      final Spec spec,
+      final SignedBeaconBlock block) {
+    this.dataAvailabilitySampler = dataAvailabilitySampler;
+    this.store = store;
+    this.kzg = kzg;
+    this.spec = spec;
+    this.block = block;
+  }
+
+  @Override
+  public boolean initiateDataAvailabilityCheck() {
+    LOG.info("Starting data availability check for slot {}", block.getSlot());
+    if (!spec.atSlot(block.getSlot())
+        .getMilestone()
+        .isGreaterThanOrEqualTo(SpecMilestone.EIP7594)) {
+      validationResult.complete(DataAndValidationResult.notRequired());
+      LOG.info("Availability check for slot {} NOT_REQUIRED, EIP7594 not started", block.getSlot());
+      return true;
+    }
+
+    if (block
+        .getBeaconBlock()
+        .orElseThrow()
+        .getBody()
+        .getOptionalBlobKzgCommitments()
+        .orElseThrow()
+        .isEmpty()) {
+      validationResult.complete(DataAndValidationResult.notRequired());
+      LOG.info(
+          "Availability check for slot {} NOT_REQUIRED, kzg commitments empty", block.getSlot());
+      return true;
+    }
+
+    final MiscHelpersEip7594 miscHelpersEip7594 =
+        MiscHelpersEip7594.required(spec.atSlot(block.getSlot()).miscHelpers());
+    if (!miscHelpersEip7594.isAvailabilityOfDataColumnSidecarsRequiredAtEpoch(
+        spec.getCurrentSlot(store), spec.computeEpochAtSlot(block.getSlot()))) {
+      validationResult.complete(DataAndValidationResult.notRequired());
+      LOG.info("Availability check for slot {} NOT_REQUIRED, epoch too old ", block.getSlot());
+      return true;
+    }
+
+    dataAvailabilitySampler
+        .checkDataAvailability(block.getSlot(), block.getRoot(), block.getParentRoot())
+        .finish(
+            dataColumnSidecars ->
+                validationResult.complete(validateImmediately(dataColumnSidecars)),
```

Comment:
Should we really validate sadecars here? I suppose that any sidecar returned by the custody should be implicitly valid